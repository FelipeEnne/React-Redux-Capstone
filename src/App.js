import React from 'react';
import './App.css';
import HeroesList from './containers/HeroesList';

const App = () => (
  <div className="App">
    <h1>Her√≥is</h1>
    <HeroesList />
  </div>
);

export default App;
