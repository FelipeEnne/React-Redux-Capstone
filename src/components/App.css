.App {
  text-align: center;
}

.app-title {
  margin-top: 10vh;
  font-size: 4rem;
}

.app-button {
  background-color: black;
  color: white;
  width: 30%;
  height: 25%;
  font-size: 3.5rem;
  font-family: 'batfont';
  margin-top: 15vh;
}
