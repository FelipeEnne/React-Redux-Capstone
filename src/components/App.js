import React, { useState, useEffect } from 'react';
import PropTypes from 'prop-types';
import './App.css';
import HeroesList from '../containers/HeroesList';
import HeroFile from './HeroFile';
// eslint-disable-next-line import/extensions
import TeamsList from '../containers/TeamsList.js';

const App = props => {
  const [heros, setHeros] = useState();
  const {
    changeRender, chngFilter, render, heroes, filter, addNewHero,
  } = props;

  useEffect(() => {
    if (!heros) {
      fetch('https://superheroapi.com/api.php/2602016920125978/search/_')
        .then(res => res.json())
        .then(json => {
          setHeros([
            json.results.filter(heroes => heroes.biography.publisher === 'DC Comics'
              || heroes.biography.publisher === 'Superman Prime One-Million'
              || heroes.biography.publisher === 'Black Racer'),
          ]);
        })
        .then(() => {
          heros['0'].forEach(hero => {
            addNewHero(hero);
          });
        });
    }
  });

  const handleAddHero = () => {
    if (heros) {
      heros.heroes.forEach(hero => {
        addNewHero(hero);
      });
    }
  };

  const handleClick = () => {
    changeRender('heroesList');
  };

  const clickHandler = (value, render) => {
    chngFilter(value);
    changeRender(render);
  };

  if (render === 'init') {
    return (
      <div className="App">
        <div id="init">
          <h1 className="app-title">Welcome to the BatComputer</h1>
          <button
            type="submit"
            onClick={handleClick}
            className="app-button"
          >
            Open subjects files
          </button>
        </div>
      </div>
    );
  }

  if (render === 'heroesList') {
    return (
      <div className="App">
        <HeroesList />
      </div>
    );
  }

  if (render === 'teamsList') {
    return (
      <div className="App">
        <TeamsList />
      </div>
    );
  }

  if (render === 'heroFile') {
    return (
      <div className="App">
        <HeroFile
          hero={heroes.filter(hero => hero.id === filter.value[1])[0]}
          filter={filter}
          handleClick={clickHandler}
        />
      </div>
    );
  }

  return null;
};


// class App extends React.Component {
//   constructor(props) {
//     super(props);
//     this.handleAddHero = this.handleAddHero.bind(this);
//     this.handleClick = this.handleClick.bind(this);
//     this.clickHandler = this.clickHandler.bind(this);
//     this.state = {
//       heroes: [],
//     };
//   }

//   componentDidMount() {
//     fetch('https://superheroapi.com/api.php/2602016920125978/search/_')
//       .then(res => res.json())
//       .then(json => {
//         this.setState({
//           heroes: json.results.filter(heroes => heroes.biography.publisher === 'DC Comics'
//             || heroes.biography.publisher === 'Superman Prime One-Million'
//             || heroes.biography.publisher === 'Black Racer'),
//         });
//       })
//       .then(this.handleAddHero);
//   }

//   handleAddHero() {
//     const { addNewHero } = this.props;
//     const { heroes } = this.state;
//     heroes.forEach(hero => {
//       addNewHero(hero);
//     });
//   }

//   handleClick() {
//     const { changeRender } = this.props;
//     changeRender('heroesList');
//   }

//   clickHandler(value, render) {
//     const { chngFilter, changeRender } = this.props;
//     chngFilter(value);
//     changeRender(render);
//   }

//   render() {
//     const { render } = this.props;

//     if (render === 'init') {
//       return (
//         <div className="App">
//           <div id="init">
//             <h1 className="app-title">Welcome to the BatComputer</h1>
//             <button
//               type="submit"
//               onClick={this.handleClick}
//               className="app-button"
//             >
//               Open subjects files
//             </button>
//           </div>
//         </div>
//       );
//     }

//     if (render === 'heroesList') {
//       return (
//         <div className="App">
//           <HeroesList />
//         </div>
//       );
//     }

//     if (render === 'teamsList') {
//       return (
//         <div className="App">
//           <TeamsList />
//         </div>
//       );
//     }

//     if (render === 'heroFile') {
//       const { heroes, filter } = this.props;
//       return (
//         <div className="App">
//           <HeroFile
//             hero={heroes.filter(hero => hero.id === filter.value[1])[0]}
//             filter={filter}
//             handleClick={this.clickHandler}
//           />
//         </div>
//       );
//     }

//     return null;
//   }
// }

App.propTypes = {
  addNewHero: PropTypes.func.isRequired,
  changeRender: PropTypes.func.isRequired,
  chngFilter: PropTypes.func.isRequired,
  render: PropTypes.string.isRequired,
  heroes: PropTypes.arrayOf(PropTypes.object).isRequired,
  filter: PropTypes.shape({
    value: PropTypes.arrayOf(PropTypes.oneOfType([
      PropTypes.string,
      PropTypes.number,
    ])),
  }).isRequired,
};

export default App;
