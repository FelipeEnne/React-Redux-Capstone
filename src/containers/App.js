import { connect } from 'react-redux';
import { addHero, chngRender } from '../actions/index';
import App from '../components/App';

const mapStateToProps = state => ({
  render: state.render,
});

const mapDispatchToProps = dispatch => ({
  addNewHero: hero => {
    dispatch(addHero(hero));
  },
  changeRender: value => {
    dispatch(chngRender(value));
  },
});

export default connect(mapStateToProps, mapDispatchToProps)(App);
