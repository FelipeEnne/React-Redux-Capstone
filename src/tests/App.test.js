import React from 'react';
import {
  render, fireEvent, screen,
} from '@testing-library/react';
import '@testing-library/jest-dom';
import App from './App';
import initialState from './initialState';

beforeEach(() => {
  delete window.location;
  window.location = new URL('http://localhost/');
});

describe('Render the page correctly', () => {
  test('Render correctly', () => {
    const { asFragment } = render(<App />);
    expect(asFragment()).toMatchSnapshot();
  });
});

describe('Open the application, click on the button e redirect to "HeroesList" page', () => {
  test('Render the application home page', () => {
    const { container } = render(<App />);
    expect(container.innerHTML).toContain('Welcome to the BatComputer');
  });

  test('Click on the button and render "HeroesList" page', () => {
    const { container } = render(<App />);
    fireEvent.click(screen.getByTestId('home-button'));
    expect(container.innerHTML).toContain('heroes-list');
    expect(container.innerHTML).toContain('Justice League of America');
    expect(container.innerHTML).not.toContain('Welcome to the BatComputer');
  });

  test('Test if the heroes data is being loaded', () => {
    const { container } = render(<App />);
    fireEvent.click(screen.getByTestId('home-button'));
    expect(container.innerHTML).toContain('Justice League of America');
    expect(container.innerHTML).toContain('<h2 class="hero-name">Superman</h2>');
    expect(container.innerHTML).toContain('<h2 class="hero-name">Wonder Woman</h2>');
    expect(container.innerHTML).toContain('<h2 class="hero-name">Batman</h2>');
  });
});

describe('Render the "HeroesList" page and click on the header`s buttons', () => {
  it('Renders the page and check the header', () => {
    const { container } = render(<App />);
    fireEvent.click(screen.getByTestId('home-button'));
    expect(container.innerHTML).toContain('Justice League of America');
  });

  it('Tests the "About" page', () => {
    const { container } = render(<App />);
    fireEvent.click(screen.getByTestId('home-button'));
    fireEvent.click(screen.getByText('About this project'));
    expect(container.innerHTML).toContain('<h1 class="about-title">About</h1>');
    expect(container.innerHTML).not.toContain('Welcome to the BatComputer');
    expect(container.innerHTML).not.toContain('Justice League of America');
  });

  it('Renders the "TeamsList" page', () => {
    const { container } = render(<App />);
    fireEvent.click(screen.getByTestId('home-button'));
    fireEvent.click(screen.getByText('Click here to change filter'));
    expect(container.innerHTML).toContain('<h1>Filter subject by alignment:</h1>');
    expect(container.innerHTML).toContain('<h1>Filter subject by filiation:</h1>');
    expect(container.innerHTML).toContain('<h1>Choose a subject by name:</h1>');
    expect(container.innerHTML).toContain('Superman');
    expect(container.innerHTML).toContain('Batman');
    expect(container.innerHTML).toContain('Wonder Woman');
    expect(container.innerHTML).not.toContain('Welcome to the BatComputer');
  });
});

describe('Render the "TeamsList" page, click on a hero and then click on a team', () => {
  it('Renders the page and clisk on a header button to render the "TeamsList" page', () => {
    const { container } = render(<App />);
    fireEvent.click(screen.getByTestId('home-button'));
    fireEvent.click(screen.getByText('Click here to change filter'));
    expect(container.innerHTML).toContain('<h1>Choose a subject by name:</h1>');
    expect(container.innerHTML).not.toContain('Welcome to the BatComputer');
  });

  initialState.heroes.forEach(hero => {
    it(`Clicks on ${hero.name}'s button and render the "HeroesFile" page`, () => {
      const { container } = render(<App />);
      fireEvent.click(screen.getByTestId('home-button'));
      fireEvent.click(screen.getByText('Click here to change filter'));
      expect(container.innerHTML).toContain(hero.name);
      fireEvent.click(screen.getByText(hero.name));
      expect(container.innerHTML).not.toContain('<h1>Choose a subject by name:</h1>');
      expect(container.innerHTML).toContain(hero.name);
      expect(container.innerHTML).toContain(hero.identity);
      expect(container.innerHTML).toContain(hero.alignment);
      expect(container.innerHTML).toContain(hero.filiation[0]);
    });
  });
});
