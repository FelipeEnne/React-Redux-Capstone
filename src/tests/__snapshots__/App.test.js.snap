// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render the page correctly Render correctly 1`] = `
<DocumentFragment>
  <div
    class="App"
    data-testid="App"
  >
    <div
      class="App"
    >
      <div
        id="init"
      >
        <h1
          class="app-title"
        >
          Welcome to the BatComputer
        </h1>
        <a
          href="/heroesList"
        >
          <button
            class="app-button"
            data-testid="home-button"
            type="submit"
          >
            Open subjects files
          </button>
        </a>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
